grpc/src/main/java/com/linecorp/armeria/server/grpc/ProtobufRequestConverterFunction.java

```diff
@@ -0,0 +1,211 @@
+ * However a <a href="https://tools.ietf.org/html/rfc7159#section-5">JSON array</a> body
+ * could be converted into a {@link Collection} type. e.g, {@code List<Message>} and {@code Set<Message>}.
+ */
+public final class ProtobufRequestConverterFunction implements RequestConverterFunction {
+
+    private static final ConcurrentMap<Class<?>, MethodHandle> methodCache =
+            new MapMaker().weakKeys().makeMap();
+    private static final Parser defaultJsonParser = JsonFormat.parser().ignoringUnknownFields();
+    private static final ObjectMapper mapper = new ObjectMapper();
+
+    private static final MethodHandle unknownMethodHandle;
+
+    static {
+        final MethodHandles.Lookup publicLookup = MethodHandles.publicLookup();
+        final MethodType mt = methodType(void.class);
+
+        MethodHandle methodHandle;
+        try {
+            methodHandle = publicLookup.findConstructor(ProtobufRequestConverterFunction.class, mt);
+        } catch (NoSuchMethodException | IllegalAccessException e) {
+            methodHandle = null;
+        }
+        assert methodHandle != null;
+        unknownMethodHandle = methodHandle;
+    }
+
+    private final ExtensionRegistry extensionRegistry;
+    private final Parser jsonParser;
+    private final ResultType resultType;
+
+    ProtobufRequestConverterFunction(ResultType resultType) {
+        jsonParser = defaultJsonParser;
+        extensionRegistry = ExtensionRegistry.getEmptyRegistry();
+        this.resultType = resultType;
+    }
+
+    /**
+     * Creates an instance with the default {@link Parser} and {@link ExtensionRegistry}.
+     */
+    public ProtobufRequestConverterFunction() {
+        this(defaultJsonParser, ExtensionRegistry.getEmptyRegistry());
+    }
+
+    /**
+     * Creates an instance with the specified {@link Parser} and {@link ExtensionRegistry}.
+     */
+    public ProtobufRequestConverterFunction(Parser jsonParser, ExtensionRegistry extensionRegistry) {
+        this.jsonParser = jsonParser;
+        this.extensionRegistry = extensionRegistry;
+        resultType = ResultType.UNKNOWN;
+    }
+
+    @Nullable
+    @Override
+    public Object convertRequest(ServiceRequestContext ctx, AggregatedHttpRequest request,
+                                 Class<?> expectedResultType,
+                                 @Nullable ParameterizedType expectedParameterizedResultType) throws Exception {
+
+        final MediaType contentType = request.contentType();
+        if (resultType == ResultType.PROTOBUF ||
+            (resultType == ResultType.UNKNOWN && Message.class.isAssignableFrom(expectedResultType))) {
+            final Message.Builder messageBuilder = getMessageBuilder(expectedResultType);
+
+            if (contentType == null ||
+                contentType.subtype().contains("protobuf") || contentType.is(MediaType.OCTET_STREAM)) {
+                return messageBuilder.mergeFrom(request.content().array(), extensionRegistry).build();
```

Comment:
Would it be better using `toInputStream()` instead of `array()` so that `HttpData` does not create a backing array?